the trait `hal::prelude::_embedded_hal_spi_FullDuplex<u8>` is not implemented for `&shared_bus_rtic::CommonBus<...`

Hello! This crate looks like exactly what I need!

I have this crate working with [radio-sx127x](https://github.com/ryankurte/rust-radio-sx127x), but I'm having trouble with [embedded-sdmmc](https://github.com/rust-embedded-community/embedded-sdmmc-rs).

If I pass `my_spi` directly instead of using `shared_spi_manager.acquire()`, embedded_sdmmc is happy, but I need the bus to be shared. Here's my (very novice) repo: https://github.com/WyseNynja/smart-compass

And here's the relevant code in `main.rs`:

```
let my_spi = hal::spi_master(
    &mut clocks,
    24.mhz(),
    device.SERCOM4,
    &mut device.PM,
    pins.sck,
    pins.mosi,
    pins.miso,
    &mut pins.port,
);

let shared_spi_manager = shared_bus_rtic::new!(my_spi, SPIMaster);

// setup the radio (works)
let radio_spi = shared_spi_manager.acquire();

let my_radio = network::Radio::new(
    radio_spi,
    rfm95_cs,
    rfm95_busy_fake,
    rfm95_int,
    rfm95_rst,
    delay,
);

// setup sd card (does NOT work)
let sd_spi = shared_spi_manager.acquire();

let sd_spi = embedded_sdmmc::SdMmcSpi::new(sd_spi, sdcard_cs);
```

```
$ cargo check
    Checking smart-compass-v1 v0.1.0 (/Users/bwstitt/code/rust-embed/smart-compass-v1)
error[E0277]: the trait bound `&shared_bus_rtic::CommonBus<hal::sercom::SPIMaster4<hal::sercom::Sercom4Pad0<hal::gpio::Pa12<hal::gpio::PfD>>, hal::sercom::Sercom4Pad2<hal::gpio::Pb10<hal::gpio::PfD>>, hal::sercom::Sercom4Pad3<hal::gpio::Pb11<hal::gpio::PfD>>>>: hal::prelude::_embedded_hal_spi_FullDuplex<u8>` is not satisfied
   --> src/main.rs:170:22
    |
170 |         let sd_spi = embedded_sdmmc::SdMmcSpi::new(sd_spi, sdcard_cs);
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `hal::prelude::_embedded_hal_spi_FullDuplex<u8>` is not implemented for `&shared_bus_rtic::CommonBus<hal::sercom::SPIMaster4<hal::sercom::Sercom4Pad0<hal::gpio::Pa12<hal::gpio::PfD>>, hal::sercom::Sercom4Pad2<hal::gpio::Pb10<hal::gpio::PfD>>, hal::sercom::Sercom4Pad3<hal::gpio::Pb11<hal::gpio::PfD>>>>`
    |
    = note: required by `embedded_sdmmc::sdmmc::SdMmcSpi::<SPI, CS>::new`

error[E0277]: the trait bound `&shared_bus_rtic::CommonBus<hal::sercom::SPIMaster4<hal::sercom::Sercom4Pad0<hal::gpio::Pa12<hal::gpio::PfD>>, hal::sercom::Sercom4Pad2<hal::gpio::Pb10<hal::gpio::PfD>>, hal::sercom::Sercom4Pad3<hal::gpio::Pb11<hal::gpio::PfD>>>>: hal::prelude::_embedded_hal_spi_FullDuplex<u8>` is not satisfied
   --> src/main.rs:170:22
    |
170 |         let sd_spi = embedded_sdmmc::SdMmcSpi::new(sd_spi, sdcard_cs);
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `hal::prelude::_embedded_hal_spi_FullDuplex<u8>` is not implemented for `&shared_bus_rtic::CommonBus<hal::sercom::SPIMaster4<hal::sercom::Sercom4Pad0<hal::gpio::Pa12<hal::gpio::PfD>>, hal::sercom::Sercom4Pad2<hal::gpio::Pb10<hal::gpio::PfD>>, hal::sercom::Sercom4Pad3<hal::gpio::Pb11<hal::gpio::PfD>>>>`
    | 
   ::: /Users/bwstitt/.cargo/registry/src/github.com-1ecc6299db9ec823/embedded-sdmmc-0.3.0/src/sdmmc.rs:20:10
    |
20  |     SPI: embedded_hal::spi::FullDuplex<u8>,
    |          --------------------------------- required by this bound in `embedded_sdmmc::sdmmc::SdMmcSpi`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `smart-compass-v1`.

To learn more, run the command again with --verbose.
```

The radio only requires `Spi: _embedded_hal_blocking_spi_Transfer<u8, Error = SpiError> + _embedded_hal_blocking_spi_Write<u8, Error = SpiError>` and not FullDuplex.

I checked `Cargo.lock` and I only have one version of embedded-hal (0.2.4).

Possibly related: I'm currently using a feather_m0 (because it's what I have), but I'm thinking it would be better to use a STM32F3DISCOVERY like they use in the rust embedded book. I don't think this is the problem though because it is working without shared-bus.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -16,3 +16,4 @@ readme = "README.md"
 
 [dependencies]
 embedded-hal = "0.2.4"
+nb = "1.0.0"

--- a/README.md
+++ b/README.md
@@ -7,7 +7,8 @@ Note that all of the drivers that use the same underlying bus **must** be stored
 resource (e.g. as one larger `struct`) within the RTIC resources. This ensures that RTIC will
 prevent one driver from interrupting another while they are using the same underlying bus.
 
-This crate also provides convenience types for working with `shared-bus` RTIC resources.
+This crate also provides convenience types for working with RTIC resources that use the same
+underlying peripheral bus.
 
 ## Usage Example
 ```rust

--- a/README.md
+++ b/README.md
@@ -28,7 +28,8 @@ struct Resources {
     shared_bus_resources: SharedBusResources<BusType>,
 }
 
-#[init] fn init(c: init::Context) -> init::LateResources {
+#[init]
+fn init(c: init::Context) -> init::LateResources {
     let manager = shared_bus_rtic::new!(bus, BusType);
     let device = Device::new(manager.acquire());
     let other_device = OtherDevice::new(manager.acquire());

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -47,7 +47,10 @@
 //! ```
 
 use core::sync::atomic::{AtomicBool, Ordering};
-use embedded_hal::blocking::{i2c, spi};
+use embedded_hal::{
+    blocking::{self, i2c},
+    spi,
+};
 
 /// A convenience type to use for declaring the underlying bus type.
 pub type SharedBus<T> = &'static CommonBus<T>;
@@ -112,21 +115,41 @@ impl<BUS: i2c::WriteRead> i2c::WriteRead for &CommonBus<BUS> {
     }
 }
 
-impl<BUS: spi::Transfer<u8>> spi::Transfer<u8> for &CommonBus<BUS> {
-    type Error = BUS::Error;
+macro_rules! spi {
+    ($($T:ty),*) => {
+        $(
+        impl<BUS: blocking::spi::Write<$T>> blocking::spi::Write<$T> for &CommonBus<BUS> {
+            type Error = BUS::Error;
 
-    fn transfer<'w>(&mut self, words: &'w mut [u8]) -> Result<&'w [u8], Self::Error> {
-        self.lock(move |bus| bus.transfer(words))
+            fn write(&mut self, words: &[$T]) -> Result<(), Self::Error> {
+                self.lock(|bus| bus.write(words))
+            }
+        }
+
+        impl<BUS: blocking::spi::Transfer<$T>> blocking::spi::Transfer<$T> for &CommonBus<BUS> {
+            type Error = BUS::Error;
+
+            fn transfer<'w>(&mut self, words: &'w mut [$T]) -> Result<&'w [$T], Self::Error> {
+                self.lock(move |bus| bus.transfer(words))
+            }
+        }
+
+        impl<BUS: spi::FullDuplex<$T>> spi::FullDuplex<$T> for &CommonBus<BUS> {
+            type Error = BUS::Error;
+
+            fn read(&mut self) -> nb::Result<$T, Self::Error> {
+                self.lock(|bus| bus.read())
+            }
+
+            fn send(&mut self, word: $T) -> nb::Result<(), Self::Error> {
+                self.lock(|bus| bus.send(word))
+            }
+        }
+        )*
     }
 }
 
-impl<BUS: spi::Write<u8>> spi::Write<u8> for &CommonBus<BUS> {
-    type Error = BUS::Error;
-
-    fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
-        self.lock(|bus| bus.write(words))
-    }
-}
+spi!(u8, u16, u32, u64);
 
 /// Provides a method of generating a shared bus.
 ///

